### ISSUE TITLE
Autoload-based automatic registration of service dependencies

### ISSUE THREAD

**Issue by jtojnar** [nette/di#282]
If I want to have some class usable for auto-wiring, it needs to be registered in the container, otherwise auto-wiring will [throw a `MissingServiceException`](https://github.com/nette/di/blob/1b2d3716c51f36f1b71780d5b0594cd0a258fa99/src/DI/Autowiring.php#L55) – I assume this is to prevent accidentally trying to create a random object in the container, which could hide that a class actually needs the object instantiated in a different way.

Currently, my options are either:

- registering the class manually by adding it into the [`services` block](https://doc.nette.org/en/dependency-injection/services), which is annoying, or
- using the [`search` extension](https://doc.nette.org/en/dependency-injection/configuration#toc-search).

The latter is based on `RobotLoader`, which, in addition to repeating the work already done by Composer in most modern projects, can cause issues when some of the classes are not loadable.

For instance, [barista](https://github.com/TomasVotruba/barista) is a Nette-based tool and it currently uses the `search` extension to locate services in the `src/` directory. But the `src/` directory also contains some helper classes, including one that depends on `PHPUnit` – even though the tool itself does not depend on PHPUnit so it might not be installed. When that is indeed the case, the search extension causes it to be loaded by [calling `class_exists`](https://github.com/nette/di/blob/1b2d3716c51f36f1b71780d5b0594cd0a258fa99/src/DI/Extensions/SearchExtension.php#L92), crashing the autoloader due to a missing PHPUnit class in the process.

Since this app is not actually needed in the first place and is only attempted to be registered due to it being in the `src/` directory, I propose adding another extension for automatic registration of dependencies in addition to `search` that, instead of a directory, would look at the dependencies of explicitly registered services and try to register them, if they are not registered already. The interface could look something like the following:

```neon
autoloaderRegistration:
    # Registers any dependency required by services that is in the Barista namespace:
    namespace: 'Barista\\'
    excludes:
      - 'Barista\\Testing\\'
```

Would that be possible to implement? My main worry would be that the container also supports runtime auto-wiring of classes that are not registered. Dependency trees of those would not be available to the compiler. But that does not seem to be the case if I am looking at [`DI\Autowiring`](https://github.com/nette/di/blob/9878f2958a0a804b08430dbc719a52e493022739/src/DI/Autowiring.php#L57) correctly. And even if it were, we could just have ignored those.

**Comment by mabar** [comment]
Solving your problem needs either ignoring files which use not autoloadable classes by configuration on your side or using static reflection and ignoring incomplete classes on Nette side.

Registering class as a service because some service relies on it is unreliable. It could work, but only with non-abstract classes. Even requiring interface instead of its implementation would break that magic. Context-specific dependencies like ORM entities would probably pass automatic registration, but who wants entities registered as services?

**Comment by dg** [comment]
I will add to SearchExtension the option to exclude files as well, I thought it was there a long time ago.

> The latter is based on RobotLoader, which, in addition to repeating the work already done by Composer in most modern projects, can cause issues when some of the classes are not loadable.

That's not true. Composer's role and capability is not to provide a list of all classes in an application.

**Comment by jtojnar** [comment]
> Composer's role and capability is not to provide a list of all classes in an application.

Yes, but I would argue most apps will know the list of classes statically so getting a list of classes only matters in two cases (AFAICT):

- A class takes an abstract class/interface as a dependency.
  - But unless there is only a single implementation, the mapping would still need to be resolved manually, thus the class name would be also known statically.
- App supports installing extensions at run time by copying them into a subdirectory.

**Comment by dg** [comment]
> But unless there is only a single implementation, the mapping would still need to be resolved manually, thus the class name would be also known statically.

And how are you going to find out that there is only one implementation and what is its name without a list of all the classes?
